Add a simple argument extractor to PackagePlugin

abertelrud opened the pull request:
This helps with simple cases until it is possible to use SwiftArgumentParser or other package dependencies from package plugins

### Motivation:

The initial version of plugins can't yet link against other modules, which excludes use of SwiftArgumentParser and similar libraries.  Plugins with more advanced needs will probably include an implementation of one of the other, simpler argument parser libraries available, but after discussion we've agreed that something minimal that operates on just string lists may make sense.  The future is clearly to use the proper argument parser, but this is a useful helper until we have that.

### Modifications:

- add an ArgumentExtractor type to the PackagePlugin module
- added a new dedicated unit test for this PackagePlugin API
- extended an end-to-end plugin unit test

### Result:

Simple extraction of argument options and flags from a plugin are possible without additional code.

abertelrud commented:
@swift-ci please smoke test

tomerd reviewed (approved):
looks good. 

missing a dedicated unit test? could come as follow up PR

abertelrud commented:
> looks good.
> 
> missing a dedicated unit test? could come as follow up PR

I changed one of the overall command plugin unit tests to test it, so we do have some coverage, but I see what you mean — it would be better to have one that we could debug through, etc, and that would add more test cases.  I'll add one in an upcoming PR.

abertelrud commented:
> > looks good.
> > missing a dedicated unit test? could come as follow up PR
> 
> I changed one of the overall command plugin unit tests to test it, so we do have some coverage, but I see what you mean — it would be better to have one that we could debug through, etc, and that would add more test cases. I'll add one in an upcoming PR.

Changed my mind on this and am adding one.  I'm waiting for the async one to land anyway, and then will have to rebase this one (probably — not sure yet) so I might as well.

abertelrud commented:
@swift-ci please smoke test

abertelrud commented on Sources/PackagePlugin/ArgumentExtractor.swift:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    /// Returns any unextracted flags or options (based strictly on whether remaining arguments have a "--" prefix).
+    public var unextractedOptionsOrFlags: [String] {
```
Should this be called `remainingOptionsOrFlags`?

abertelrud commented:
@swift-ci please smoke test
